stage2.1_traited_script/traited_face_detect.py

```diff
@@ -0,0 +1,90 @@
+""" Script analyzing an image, detecting human faces inside it, and printing
+EXIF data about it.
+"""
+import PIL.Image
+from PIL.ExifTags import TAGS
+from skimage import data
+from skimage.feature import Cascade
+import matplotlib.pyplot as plt
+from matplotlib import patches
+from os.path import join
+import numpy as np
+
+# ETS imports
+from traits.api import Dict, File, HasStrictTraits, List, observe
+
+
+class ImageFile(HasStrictTraits):
+    """ Model to hold an image file.
+    """
+    filepath = File
+
+    faces = List
+
+    metadata = Dict
+
+    def to_array(self):
+        with PIL.Image.open(self.filepath) as img:
+            return np.asarray(img)
+
+    @observe("filepath")
+    def update_metadata(self, event):
+
+        with PIL.Image.open(self.filepath) as img:
+            exif = img._getexif()
+        self.metadata = {TAGS[k]: v for k, v in exif.items()
+                         if k in TAGS}
+
+    def detect_faces(self):
+        # Load the trained file from the module root.
+        trained_file = data.lbp_frontal_face_cascade_filename()
+
+        # Initialize the detector cascade.
+        detector = Cascade(trained_file)
+
+        detected = detector.detect_multi_scale(img=self.to_array(),
+                                               scale_factor=1.2,
+                                               step_ratio=1,
+                                               min_size=(60, 60),
+                                               max_size=(600, 600))
+        self.faces = detected
+
+        self.metadata["Number of faces detected"] = len(detected)
+
+
+# Select image file -----------------------------------------------------------
+
+image_path = join("..", "sample_images", "IMG-0311_xmas_2020.JPG")
+image_path2 = join("..", "sample_images", "owls.jpg")
+
+img = ImageFile()
+for path in [image_path, image_path2]:
+
+    img.filepath = path
+
+    # Detect faces ------------------------------------------------------------
+
+    img.detect_faces()
+
+    print(img.metadata)
+
+    # Visualize results -------------------------------------------------------
+
+    plt.imshow(img.to_array())
```

Comment:
Should there be a method on ImageFile, that does the plotting? That way the main code would just be 3-4 lines?

Reply:
I would propose to keep moving plotting/output to a class for the stage 3 when we bring in views. OK with you?

Reply:
> I would propose to keep moving plotting/output to a class for the stage 3 when we bring in views. 

That would be my vote as well. The plotting folds smoothly into the view for `ImageFile` (see #44, e.g.) and so a heavy refactor of the script doesn't seem necessary.

Reply:
I still feel that having plotting code in the main namespace is a big "smell" to me. I would think that the minimum refactor should really require that the code be clean enough that it be importable from an IPython console session.   I would prefer a clean OO structure at first, a reasonable first cut redesign is to put the plotting into the model. Then we argue that it is better to refactor into a separate view since the plotting has nothing per se to do with the "model". Although one could argue that the model often becomes trivial in many cases at which point a data class and a separate view for it to me seems like overkill.  I think there is a fine balance between too many classes and too few but we don't need to go there.

Reply:
This is a really great conversation, thanks guys! I think @siddhantwahal  and I have the same idea as to what path to take attendees, but I see where you are coming from @prabhuramachandran . You are basically wanting to OOP the script first, and them make the model and the view emerge from a single class. I am ok with that. Also, this kinda echos a suggestion of mine on stage 3 to start with a `configure_traits` call on the model (or unique class) for 3.0, add a `traits_view` method on the class (3.1) and then split to MVC (3.2). Prabhu is coming at it from the point of view of keeping simple to not lose too many people too early, and I (and @siddhantwahal ?) am wanting to force a few patterns down their throat as part of my desire to teach them building scalable solutions. These patterns might seem overkill, and obviously are for a toy example like here, but are life saviors IMO if you want to build something non-trivial. If we all agree on these 2 goals, we can continue to argue as to when to introduce what.

Reply:
@jonathanrocher -- thanks for being patient with my suggestions. I agree that a good structure is important but that too should come in as a stage, each jump should be small IMO.  I think terrible throwaway code -> cleaner OO code -> traits model code -> simple traitsui view -> developer sold on overall idea.  Stop. Now think about scalability, introduce a general pattern and layout -> Refactor old code to new layout -> Cleanup, more features -> Package it with setup.py -> Build installer -> ship it!